File: Common-Interview-Problems/MaximumSubarray.py
# Given an array of integers, 
# find the contiguous subarray (containing at least one number) 
# which has the largest sum and return its sum.

# Kadanes Algorithm: O(N) time complexity 
def kadanes_algorithm(arr):
    max_global = arr[0]
    max_current = arr[0]
    for i in range(1, len(arr)):
        max_current = max(arr[i], max_current + arr[i])
        if max_current > max_global:
            max_global = max_current
    return max_global

File: Common-Interview-Problems/MaximumSubarray.js
/*
Given an array of integers, 
find the contiguous subarray (containing at least one number) 
which has the largest sum and return its sum.
*/

// Kadanes Algorithm: O(N) time complexity O(1) space
function kadanesAlgorithm(arr) {
    let currentMax = arr[0]
    let globalMax = arr[0]
    for (let i = 1; i < arr.length; i++) {
        currentMax = Math.max(arr[i], currentMax + arr[i])
        if (currentMax > globalMax) {
            globalMax = currentMax
        }
    }
    return globalMax
}
